integration/gethnetwork/geth_network.go

```diff
@@ -0,0 +1,172 @@
+package gethnetwork
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+const (
+	genesisFileName = "genesis.json"
+	nodeFolderName  = string(os.PathSeparator) + "node%d_datadir" + string(os.PathSeparator)
+	ipcFileName     = "geth.ipc"
+	tempDirPrefix   = "geth_nodes"
+
+	addPeerCmd = "admin.addPeer(%s)"
+	attachCmd  = "attach"
+	enodeCmd   = "admin.nodeInfo.enode"
+	initCmd    = "init"
+
+	dataDirFlag      = "--datadir"
+	execFlag         = "--exec"
+	portFlag         = "--port"
+	mineFlag         = "--mine"
+	minerThreadsFlag = "--miner.threads=1"
+	minerEthbaseFlag = "--miner.etherbase=0x0000000000000000000000000000000000000001"
+
+	// We pre-allocate a single wallet, which is shared by all nodes.
+	genesisConfig = `{
+	  "config": {
+		"chainId": 777,
+		"homesteadBlock": 0,
+		"eip150Block": 0,
+		"eip155Block": 0,
+		"eip158Block": 0,
+		"byzantiumBlock": 0,
+		"constantinopleBlock": 0,
+		"petersburgBlock": 0,
+		"istanbulBlock": 0,
+		"berlinBlock": 0,
+		"londonBlock": 0
+	  },
+	  "alloc": {
+		"0x0000000000000000000000000000000000000001": {
+		  "balance": "111111111"
+		}
+	  },
+	  "coinbase": "0x0000000000000000000000000000000000000000",
+	  "difficulty": "0x20000",
+	  "extraData": "",
+	  "gasLimit": "0x2fefd8",
+	  "nonce": "0x0000000000000042",
+	  "mixhash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+	  "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+	  "timestamp": "0x00"
+  }`
+)
+
+// GethNetwork is a network of Geth nodes, built using the provided Geth binary.
+type GethNetwork struct {
+	gethBinaryPath  string
+	genesisFilePath string
+	dataDirs        []string
+}
+
+// NewGethNetwork using the provided Geth binary to create a private Ethereum network with numNodes Geth nodes.
+func NewGethNetwork(gethBinaryPath string, numNodes int) GethNetwork {
+	nodesDir, err := ioutil.TempDir("", tempDirPrefix)
+	if err != nil {
+		panic(err)
+	}
+
+	// We write out the `genesis.json` file to be used by the network.
+	genesisFilePath := nodesDir + string(os.PathSeparator) + genesisFileName
+	err = os.WriteFile(genesisFilePath, []byte(genesisConfig), 0o600)
+	if err != nil {
+		panic(err)
+	}
+
+	// Each Geth node needs its own data directory.
+	dataDirs := make([]string, numNodes)
+	for i := 0; i < numNodes; i++ {
+		dataDirs[i] = fmt.Sprintf(nodesDir+nodeFolderName, i+1)
+	}
+
+	network := GethNetwork{
+		gethBinaryPath:  gethBinaryPath,
+		genesisFilePath: genesisFilePath,
+		dataDirs:        dataDirs,
```

Comment:
Can we have a comment explaining why we have to init and then start please ?

Reply:
Added.